lib/Dialect/TritonGPU/Transforms/OptimizeDotOperands.cpp

```diff
@@ -213,7 +213,15 @@ class UseShmemForScales
     }
 
     auto localLoad = getNextOp<triton::gpu::LocalLoadOp>(reshapeOp5D.getSrc());
-    if (!localLoad || !isTmemCopyCompatible(localLoad.getSrc().getType())) {
+    if (localLoad) {
+      auto localAlloc = getNextOp<LocalAllocOp>(localLoad.getSrc());
+      bool usesTMAload = false;
+      if (localAlloc && localAlloc.getSrc())
+        usesTMAload =
+            getNextOp<DescriptorLoadOp>(localAlloc.getSrc()) != nullptr;
+      if (!isTmemCopyCompatible(localLoad.getSrc().getType(), usesTMAload))
+        return failure();
+    } else {
       return failure();
     }
```

Comment:
NIT: We can flatten out some of the nesting
```suggestion
    if (!localLoad) {
      return failure();
    }
    auto localAlloc = getNextOp<LocalAllocOp>(localLoad.getSrc());
    bool usesTMAload = (
        localAlloc && 
        localAlloc.getSrc() &&
        (getNextOp<DescriptorLoadOp>(localAlloc.getSrc()) != nullptr)
    );
    if (!isTmemCopyCompatible(localLoad.getSrc().getType(), usesTMAload))
      return failure();
```